src/otx/engine/engine.py

```diff
@@ -0,0 +1,407 @@
+"""Module for OTX engine components."""
+
+# Copyright (C) 2023 Intel Corporation
+# SPDX-License-Identifier: Apache-2.0
+
+from __future__ import annotations
+
+from typing import TYPE_CHECKING, Any, Iterable
+
+import torch
+from lightning import Trainer, seed_everything
+
+from otx.core.config.device import DeviceConfig
+from otx.core.data.module import OTXDataModule
+from otx.core.model.entity.base import OTXModel
+from otx.core.model.module.base import OTXLitModule
+from otx.core.types.device import DeviceType
+from otx.core.types.task import OTXTaskType
+from otx.core.utils.cache import TrainerArgumentsCache
+
+if TYPE_CHECKING:
+    from pathlib import Path
+
+    from lightning import Callback
+    from lightning.pytorch.cli import LRSchedulerCallable, OptimizerCallable
+    from lightning.pytorch.loggers import Logger
+    from lightning.pytorch.utilities.types import EVAL_DATALOADERS
+    from pytorch_lightning.trainer.connectors.accelerator_connector import _PRECISION_INPUT
+
+
+LITMODULE_PER_TASK = {
+    OTXTaskType.MULTI_CLASS_CLS: "otx.core.model.module.classification.OTXMulticlassClsLitModule",
+    OTXTaskType.MULTI_LABEL_CLS: "otx.core.model.module.classification.OTXMultilabelClsLitModule",
+    OTXTaskType.H_LABEL_CLS: "otx.core.model.module.classification.OTXHlabelClsLitModule",
+    OTXTaskType.DETECTION: "otx.core.model.module.detection.OTXDetectionLitModule",
+    OTXTaskType.INSTANCE_SEGMENTATION: "otx.core.model.module.instance_segmentation.OTXInstanceSegLitModule",
+    OTXTaskType.SEMANTIC_SEGMENTATION: "otx.core.model.module.segmentation.OTXSegmentationLitModule",
+    OTXTaskType.ACTION_CLASSIFICATION: "otx.core.model.module.action_classification.OTXActionClsLitModule",
+    OTXTaskType.ACTION_DETECTION: "otx.core.model.module.action_detection.OTXActionDetLitModule",
+}
+
+
+class Engine:
+    """OTX Engine.
+
+    This class defines the common interface for OTX, including methods for training and testing.
+
+    Example:
+        # Auto-Configuration with data_root
+        >>> engine = Engine(
+            data_root=<dataset/path>,
+        )
+        # With Custom OTXModel
+        >>> engine = Engine(
+            data_root=<dataset/path>,
+            model=OTXModel(...),
+            checkpoint=<checkpoint/path>,
+        )
+        # With Custom OTXDataModule
+        >>> engine = Engine(
+            model = OTXModel(...),
+            datamodule = OTXDataModule(...),
+        )
```

Comment:
When there are multiple examples, the section name becomes Examples in sphinx. We also need to use `...` when we break the python code multiple lines.

```suggestion
    Examples:
    The following examples show how to use the Engine class.

        .. code-block:: python
            :caption: Auto-Configuration with data_root

            engine = Engine(data_root=<dataset/path>)

        .. code-block:: python
            :caption: With Custom OTXModel

            engine = Engine(
               data_root=<dataset/path>,
               model=OTXModel(...),
               checkpoint=<checkpoint/path>,
            )

        .. code-block:: python
            :caption: With Custom OTXDataModule

            engine = Engine(
               model = OTXModel(...),
               datamodule = OTXDataModule(...),
            )
```

Reply:
or if you think that code block is too long, you could replace it with 

```python
            >>> engine = Engine(
            ...  model = OTXModel(...),
            ...   datamodule = OTXDataModule(...),
            ... )
```
we need to add `...` when multiple lines exist

Reply:
Thanks! Done with 2nd way - https://github.com/openvinotoolkit/training_extensions/pull/2778/commits/367d9bc4dcbb549d9c065e7286abf587d1a36675